rpc/grpc/server/services/pruningservice/service.go

```diff
@@ -0,0 +1,89 @@
+package pruningservice
+
+import (
+	context "context"
+	"fmt"
+	"math"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/cometbft/cometbft/libs/log"
+	v1 "github.com/cometbft/cometbft/proto/tendermint/services/pruning/v1"
+	sm "github.com/cometbft/cometbft/state"
+)
+
+type pruningServiceServer struct {
+	pruner *sm.Pruner
+	logger log.Logger
+}
+
+// New creates a new CometBFT pruning service server.
+func New(pruner *sm.Pruner, logger log.Logger) v1.PruningServiceServer {
+	return &pruningServiceServer{
+		pruner: pruner,
+		logger: logger.With("service", "PruningService"),
+	}
+}
+
+// SetBlockRetainHeight implements v1.PruningServiceServer.
+func (s *pruningServiceServer) SetBlockRetainHeight(_ context.Context, req *v1.SetBlockRetainHeightRequest) (*v1.SetBlockRetainHeightResponse, error) {
+	height := req.Height
+	// Because we can't agree on a single type to represent block height.
+	if height > uint64(math.MaxInt64) {
```

Comment:
Height should also never be < 0 .

Reply:
It is by virtue of its `uint64` encoding in protobuf.